documentation/src/docker.md

```diff
@@ -0,0 +1,108 @@
+# Forest in Docker🌲❤️🐋
+
+## Prerequisites
+- Docker engine installed and running. Forest containers are confirmed to run on the following engines:
+  - Docker Engine (Community) on Linux,
+  - Docker for macOS
+  - Podman on WSL
+
+Native images are available for the following platforms:
+- `linux/arm64`
+- `linux/amd64`
+
+The images will work out-of-the box on both Intel processors and macOS with M1/M2.
+
+## Tags
+For the list of all available tags please refer to the [Forest packages](https://github.com/ChainSafe/forest/pkgs/container/forest).
+
+Currently, the following tags are produced:
+- `latest` - latest build of the `main` branch.
+- `date-digest` e.g., `2023-02-17-5f27a62` - all builds that landed on the `main` branch.
+- release tags, available from `v.0.7.0` onwards.
+
+## Usage
+
+### List available flags and/or commands
+```shell
+# daemon
+❯ docker run --init -it --rm ghcr.io/chainsafe/forest:latest --help
+# cli
+❯ docker run --init -it --rm --entrypoint forest-cli ghcr.io/chainsafe/forest:latest --help
+```
+
+### Create a Forest node running calibration network. Then list all connected peers.
+
+```shell
+❯ docker run --init -it --rm --name forest ghcr.io/chainsafe/forest:latest --chain calibnet --auto-download-snapshot
+```
+then in another terminal (sample output)
+```shell
+❯ docker exec -it forest forest-cli net peers
+12D3KooWAh4qiT3ZRZgctVJ8AWwRva9AncjMRVBSkFwNjTx3EpEr, [/ip4/10.0.2.215/tcp/1347, /ip4/52.12.185.166/tcp/1347]
+12D3KooWMY4VdMsdbFwkHv9HxX2jZsUdCcWFX5F5VGzBPZkdxyVr, [/ip4/162.219.87.149/tcp/30141, /ip4/162.219.87.149/tcp/30141/p2p/12D3KooWMY4VdMsdbFwkHv9HxX2jZsUdCcWFX5F5VGzBPZkdxyVr]
+12D3KooWFWUqE9jgXvcKHWieYs9nhyp6NF4ftwLGAHm4sCv73jjK, [/dns4/bootstrap-3.calibration.fildev.network/tcp/1347]
+```
+
+### Use a shared volume to utilise across different Forest images
+Create the volume
+```shell
+docker volume create forest-data
+```
+Now, whenever you create a new Forest container, attach the volume to where the data is stored `/home/forest/.local/share/forest`.
+```shell
+❯ docker run --init -it --rm --network host \
```

Comment:
Better to set ulimit as well by `--ulimit nofile=60000`

Reply:
`--network host` looks scary.  Better to map all ports or we could at least suggest using rootless docker that mitigates this vulnerability.

https://docs.docker.com/engine/security/rootless/
> Host network (docker run --net=host) is also namespaced inside RootlessKit.

Reply:
As per my testing result, setting `--memory-swappiness 1` is beneficial, however, the parameter does not seem to work with the default docker installation, alternatively we could ask setting `sudo sysctl -w vm.swappiness=1` on host machine from which docker container will inherit.
For more details: https://github.com/ChainSafe/forest/issues/2563

Reply:
> `--network host` looks scary. Better to map all ports or we could at least suggest using rootless docker that mitigates this vulnerability.
> 
> https://docs.docker.com/engine/security/rootless/
> 
> > Host network (docker run --net=host) is also namespaced inside RootlessKit.

It's scary because it shouldn't be here. I guess I mindlessly copied it somewhere from the depths of my shell history. Good catch.

Reply:
@hanabi1224 As per our prior GH and Slack conversation, could you create a section here with Docker setup recommendations? Your points regarding running rootless are more than valid, and mentioning all those things (podman included) would be beneficial. It'd be a shame if it ended up as another forgotten Slack thread. Mentioning the swappiness would also make sense there, given it's not necessarily required but highly recommended.

Reply:
> 60000

Why this value exactly?

Reply:
> > 60000
> 
> Why this value exactly?

I just randomly chose a big value

Reply:
> > `--network host` looks scary. Better to map all ports or we could at least suggest using rootless docker that mitigates this vulnerability.
> > https://docs.docker.com/engine/security/rootless/
> > > Host network (docker run --net=host) is also namespaced inside RootlessKit.
> 
> It's scary because it shouldn't be here. I guess I mindlessly copied it somewhere from the depths of my shell history. Good catch.

Note that without mapping the libp2p port, the listen addr does not work, all connections to other nodes have to be initiated from the forest daemon

Reply:
> Note that without mapping the libp2p port, the listen addr does not work, all connections to other nodes have to be initiated from the forest daemon

That's not great. What is the practical impact of that? The host network doesn't really work great on Docker for Mac as far as I remember.

Perhaps using a dedicated port for libp2p is the way to go? What do you think?

Reply:
> @hanabi1224 As per our prior GH and Slack conversation, could you create a section here with Docker setup recommendations? Your points regarding running rootless are more than valid, and mentioning all those things (podman included) would be beneficial. It'd be a shame if it ended up as another forgotten Slack thread. Mentioning the swappiness would also make sense there, given it's not necessarily required but highly recommended.


### Security recommendations
- We strongly recommend running the docker daemon in rootless mode ([installation instruction](https://docs.docker.com/engine/security/rootless/)), or running the daemon-less docker alternative `podman` ([installation instruction](https://podman.io/getting-started/installation)) with non-root user and put `alias docker = podman` or manually replace the `docker` commands with `podman` in below instructions

### Performance recommendations
- We recommend lowering the swappiness kernel parameter on linux to 1-10 for long running forest node by doing `sudo sysctl -w vm.swappiness=[n]`. References: [[1]](https://en.wikipedia.org/wiki/Memory_paging#Swappiness)  [[2]](https://linuxhint.com/understanding_vm_swappiness/)

Reply:
> > Note that without mapping the libp2p port, the listen addr does not work, all connections to other nodes have to be initiated from the forest daemon
> 
> That's not great. What is the practical impact of that? The host network doesn't really work great on Docker for Mac as far as I remember.
> 
> Perhaps using a dedicated port for libp2p is the way to go? What do you think?

We could track this in a separate issue. With the current bootstrap peer list and the max peers config, the impact of not listening to an public address is trival

Reply:
Thanks, co-authored the recommendations.